Unable to query back/foreground color with gnome-terminal/tilix (libvte)

Hi,

thanks for this awesome library! I've discovered that with the current code, I'm unable to query the foreground/background color in terminals using libvte, which includes gnome-terminal and tilix on Debian stable (libvte 0.54.2). It works with urxvt though.

I've looked into the issue, building the `examples/hello-world` and running it with strace reveals what's happening (using tilix with a light background color, `#fafafa`):

```
$ cd examples/hello-world
$ go build
$ strace -e read,write -s 999 ./hello-world
[...]
write(1, "\33]11;?\7", 7)               = 7
read(1, "\33", 1)                       = 1
read(1, "]", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, ";", 1)                         = 1
read(1, "r", 1)                         = 1
read(1, "g", 1)                         = 1
read(1, "b", 1)                         = 1
read(1, ":", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "/", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "/", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "f", 1)                         = 1
read(1, "a", 1)                         = 1
read(1, "\33", 1)                       = 1
read(1, "\\", 1)                        = 1
write(1, "\n\t\33[1mHas dark background?\33[0m true\n", 36
	Has dark background? true
) = 36
+++ exited with 0 +++
```

We can see here that termenv sends the query to the terminal using a `BEL` character as the terminator (`"\33]11;?\7"`). Then tilix/libvte responds with the color, terminated by `ESC \`, which is rejected by termenv.

On urxvt, the response is terminated by `BEL` (with dark background `#111111` this time):
```
$ strace -e read,write -s 999 ./hello-world
[...]
write(1, "\33]11;?\7", 7)               = 7
read(1, "\33", 1)                       = 1
read(1, "]", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, ";", 1)                         = 1
read(1, "r", 1)                         = 1
read(1, "g", 1)                         = 1
read(1, "b", 1)                         = 1
read(1, ":", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "/", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "/", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "1", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "0", 1)                         = 1
read(1, "\7", 1)                        = 1
[...]
```

It looks like a bug in libvte 0.54.4. I've also tested this on Fedora 32 using libvte 0.60.3, there it is fixed. I was also able to find a warning in the [source code](https://gitlab.gnome.org/GNOME/vte/-/blob/master/src/parser-cat.cc#L736-739):

```
if (seq.st() == 7 /* BEL */)
    warn("OSC terminated by BEL may be ignored; use ST (ESC \\) instead.");
```

According to [the standard](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html) I was able to find, OSC may be terminated by either `BEL` or `ESC \` (named `ST`):

> Operating System Commands
>
> OSC Ps ; Pt BEL
> OSC Ps ; Pt ST

I suspect that there are many more terminal emulators which just support `ESC \` as terminator. It feels to me they all just copied code from xterm at some point...

I can imagine solving this issue in two ways:
 * Using `ESC \` instead of `BEL` when querying the colors, checking that the response is also terminated by `ESC \`.
 * Using `BEL `for querying the colors, but accepting `BEL` as well as `ESC \` as the response terminator.

If either solution would be acceptable, I'm willing to submit a PR. Please let me know! :)

PR is submitted. BTW, I noticed that running the tests interactively (i.e. with a real terminal) fails:
```
$ --- FAIL: TestTermEnv (0.00s)
    termenv_test.go:23: Expected 97, got 38;2;105;113;121
    termenv_test.go:30: Expected 48;2;0;0;0, got 48;2;250;250;250
FAIL
exit status 1
FAIL	github.com/muesli/termenv	0.003s
```

Running it through `cat` (so that `stdout` is not directly connected to the terminal) makes them pass:
```
$ go test | cat
PASS
ok  	github.com/muesli/termenv	0.002s
```

(current `master` branch, tested on tilix and alacritty)

--- a/color.go
+++ b/color.go
@@ -153,9 +153,19 @@ func (c RGBColor) Sequence(bg bool) string {
 }
 
 func xTermColor(s string) (RGBColor, error) {
-	if len(s) != 24 {
+	if len(s) < 24 || len(s) > 25 {
 		return RGBColor(""), ErrInvalidColor
 	}
+
+	switch {
+	case strings.HasSuffix(s, "\a"):
+		s = strings.TrimSuffix(s, "\a")
+	case strings.HasSuffix(s, "\033\\"):
+		s = strings.TrimSuffix(s, "\033\\")
+	default:
+		return RGBColor(""), ErrInvalidColor
+	}
+
 	s = s[4:]
 
 	prefix := ";rgb:"
@@ -163,7 +173,6 @@ func xTermColor(s string) (RGBColor, error) {
 		return RGBColor(""), ErrInvalidColor
 	}
 	s = strings.TrimPrefix(s, prefix)
-	s = strings.TrimSuffix(s, "\a")
 
 	h := strings.Split(s, "/")
 	hex := fmt.Sprintf("#%s%s%s", h[0][:2], h[1][:2], h[2][:2])

--- a/color_test.go
+++ b/color_test.go
@@ -1,0 +1,59 @@
+package termenv
+
+import "testing"
+
+func TestXTermColor(t *testing.T) {
+	var tests = []struct {
+		input string
+		color RGBColor
+		valid bool
+	}{
+		{
+			"\033]11;rgb:fafa/fafa/fafa\033\\",
+			RGBColor("#fafafa"),
+			true,
+		},
+		{
+			"\033]11;rgb:1212/3434/5656\a",
+			RGBColor("#123456"),
+			true,
+		},
+		{
+			"\033]11;foo:fafa/fafa/fafaZZ",
+			"",
+			false,
+		},
+		{
+			"\033]11;rgb:fafa/fafa",
+			"",
+			false,
+		},
+		{
+			"\033]11;rgb:fafa/fafa/fafaY",
+			"",
+			false,
+		},
+		{
+			"\033]11;rgb:fafa/fafa/fafaZZ",
+			"",
+			false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run("", func(t *testing.T) {
+			color, err := xTermColor(test.input)
+			if err != nil && test.valid {
+				t.Fatalf("unexpected error for input %q: %v", test.input, err)
+			}
+
+			if err == nil && !test.valid {
+				t.Fatalf("expected error for input %v not found", test.input)
+			}
+
+			if color != test.color {
+				t.Fatalf("wrong color returned, want %v, got %v", test.color, color)
+			}
+		})
+	}
+}

--- a/termenv_unix.go
+++ b/termenv_unix.go
@@ -3,6 +3,7 @@
 package termenv
 
 import (
+	"bytes"
 	"fmt"
 	"os"
 	"runtime"
@@ -121,7 +122,9 @@ func readWithTimeout(f *os.File) (string, bool) {
 		// log.Printf("read %d bytes from stdout: %s %d\n", n, data, data[len(data)-1])
 
 		data = append(data, b[0])
-		if b[0] == '\a' || (b[0] == '\\' && len(data) > 2) {
+
+		// data sent by terminal is either terminated by BEL (\a) or ST (ESC \)
+		if bytes.HasSuffix(data, []byte("\a")) || bytes.HasSuffix(data, []byte("\033\\")) {
 			break
 		}
 	}
@@ -144,7 +147,7 @@ func termStatusReport(sequence int) (string, error) {
 		return "", ErrStatusReport
 	}
 
-	fmt.Printf("\033]%d;?\007", sequence)
+	fmt.Printf("\033]%d;?\033\\", sequence)
 	s, ok := readWithTimeout(os.Stdout)
 	if !ok {
 		return "", ErrStatusReport

